enginetest/histogram_test.go

```diff
@@ -0,0 +1,514 @@
+	runSuite(t, statTests, 100, 20, debug)
+	runSuite(t, statTests, 500, 10, debug)
+	runSuite(t, statTests, 500, 20, debug)
+}
+
+func TestMultiDist(t *testing.T) {
+	tests := []statsTest{
+		{
+			name: "uniform dist int",
+			tableGen: func(ctx *sql.Context, db *memory.Database, cnt int, tab sql.TableId, col sql.ColumnId, args ...interface{}) *plan.ResolvedTable {
+				xyz := makeTable(db, "xyz", tab, col)
+				err := uniformDistForTable(ctx, xyz, cnt)
+				if err != nil {
+					panic(err)
+				}
+				return xyz
+			},
+			leftOrd:    []int{1},
+			rightOrd:   []int{1},
+			leftTypes:  []sql.Type{types.Int64, types.Int64, types.Int64},
+			rightTypes: []sql.Type{types.Int64, types.Int64, types.Int64},
+			err:        .1,
+		},
+	}
+
+	runSuite(t, tests, 1000, 10, false)
+}
+
+func runSuite(t *testing.T, tests []statsTest, rowCnt, bucketCnt int, debug bool) {
+	for i, tt := range tests {
+		t.Run(fmt.Sprintf("%s: , rows: %d, buckets: %d", tt.name, rowCnt, bucketCnt), func(t *testing.T) {
+			db := memory.NewDatabase(fmt.Sprintf("test%d", i))
+			pro := memory.NewDBProvider(db)
+
+			xyz := tt.tableGen(newContext(pro), db, rowCnt, sql.TableId(i*2), 1, tt.args1...)
+			wuv := tt.tableGen(newContext(pro), db, rowCnt, sql.TableId(i*2+1), sql.ColumnId(len(tt.leftTypes)+1), tt.args2...)
+
+			// join the histograms on a particular field
+			var joinOps []sql.Expression
+			for i, l := range tt.leftOrd {
+				r := tt.rightOrd[i]
+				joinOps = append(joinOps, eq(l, r+len(tt.leftTypes)))
+			}
+
+			exp, err := expectedResultSize(newContext(pro), xyz, wuv, joinOps, debug)
+			require.NoError(t, err)
+
+			// get histograms for tables
+			xHist, err := testHistogram(newContext(pro), xyz, tt.leftOrd, bucketCnt)
+			require.NoError(t, err)
+
+			wHist, err := testHistogram(newContext(pro), wuv, tt.rightOrd, bucketCnt)
+			require.NoError(t, err)
+
+			if debug {
+				log.Printf("xyz:\n%s\n", sql.Histogram(xHist).DebugString())
+				log.Printf("wuv:\n%s\n", sql.Histogram(wHist).DebugString())
+			}
+
+			lStat := &stats.Statistic{Typs: []sql.Type{types.Int64}}
+			for _, b := range xHist {
+				lStat.Hist = append(lStat.Hist, b.(*stats.Bucket))
+			}
+			rStat := &stats.Statistic{Typs: []sql.Type{types.Int64}}
+			for _, b := range wHist {
+				rStat.Hist = append(rStat.Hist, b.(*stats.Bucket))
+			}
+
+			res, err := stats.Join(stats.UpdateCounts(lStat), stats.UpdateCounts(rStat), []int{0}, []int{0}, debug)
+			require.NoError(t, err)
+			if debug {
+				log.Printf("join %s\n", res.Histogram().DebugString())
+			}
+
+			denom := float64(exp)
+			if cmp := float64(res.RowCount()); cmp > denom {
+				denom = cmp
+			}
+			delta := float64(exp-int(res.RowCount())) / denom
+			if delta < 0 {
+				delta = -delta
+			}
+			if debug {
+				log.Println(res.RowCount(), exp, delta)
+			}
+			require.Less(t, delta, tt.err, "%d/%d/%.2f\nleft %s\nright %s", res.RowCount(), exp, delta, sql.Histogram(xHist).DebugString(), sql.Histogram(wHist).DebugString())
```

Comment:
Can you add a comment explaining what specifically is being checked on this line?